add distance decay access options

shiqin-liu opened the pull request:
This is to address #111 on distance decay function for accessibility measures, the change include

1. add the three accessibility methods functions in setup_sp.py to include binary, soft threshold, and Cumulative–Gaussian functions
2. update sp.py with binary function 

Updates:
- [x] provide default values of k for a given default access distance threshold (Resolved [comment](https://github.com/global-healthy-liveable-cities/global-indicators/pull/112#discussion_r580677870))
- [x] simplify accessibility score functions without the new names (Resolved [comment](https://github.com/global-healthy-liveable-cities/global-indicators/pull/112#discussion_r580677870))
- [x] doing fillna(0) prior to casting type to avoid the need of using int64 for a nullable datatype. (Resolved [comment](https://github.com/global-healthy-liveable-cities/global-indicators/pull/112#discussion_r580607360))
- [x] update cumulative–gaussian function, without the need of for loop (Resolved [comment](https://github.com/global-healthy-liveable-cities/global-indicators/pull/112#discussion_r580695208))
- [x] amend the variable names and suffixes relating to the word 'binary' in the implementation code ([comment](https://github.com/global-healthy-liveable-cities/global-indicators/pull/112#discussion_r580663297))

@carlhiggs feel free to do another PR for the configuration setup as you see fits :)

@gboeing  I think this PR should be sufficient to address the reviewers' comment,  all three functions can run properly, so should be good to merge

shiqin-liu committed:
add distance decay access options

gboeing commented on process/setup_sp.py:
```diff
@@ -32,7 +33,7 @@ def read_proj_graphml(proj_graphml_filepath, ori_graphml_filepath, to_crs,undire
     retain_edge_attributes = list (default: None)
         explicitly retain only a subset of edge attributes, else keep all (default)
-     
+
```
I'm noticing that probably 1/2 of the diffs here are whitespace changes. I wonder if we should enforce code guidelines and formatting (like black) to keep things more consistent.

carlhiggs replied:
I think that's a good idea.  Regarding code guidelines and formatting, we could try using a code linter to guide style enforcement perhaps @shiqin-liu @gboeing ?  I haven't used one before but I've been meaning to start trying to make a habit of it; sounds like it might be appropriate for this issue.  This [blog post]( https://dbader.org/blog/python-code-linting ) recommends Flake8 for conformity to PEP8 style. Maybe either that or PyLint could be useful ([reddit thread here](https://www.reddit.com/r/Python/comments/82hgzm/any_advantages_of_flake8_over_pylint/))?  Would you have any recommendations or thoughts around this Shirley or Geoff?

gboeing replied:
I'd recommend a setup similar to that of OSMnx: isort + black + flake8 + pydocstyle for linting. And then build/unit tests as needed. OSMnx linting and testing is [here](https://github.com/gboeing/osmnx/tree/master/tests) and some [configs](https://github.com/gboeing/osmnx/blob/master/setup.cfg) are by necessity at the project root.

gboeing replied:
One other comment: this all uses continuous integration, which is configured [here](https://github.com/gboeing/osmnx/blob/master/.github/workflows/tests.yml) as an example.

carlhiggs replied:
I like the sounds of it -- although I wouldn't be sure how to get started with it all; I'll try to look into this later in the week so I'm familiar with the process at least; maybe we can talk about possible implementation in our meeting later in the week?

shiqin-liu replied:
Thanks Geoff. I noticed the space change but had no idea why this happened... It sounds a great idea to enforce code guidelines and formatting to deal with this. This is quite new to me, I will aim to take a look at the examples and guides. I would appreciate if someone could lend a hand too :)

gboeing replied:
I can take the lead getting continuous integration and linting set up.

gboeing commented on process/setup_sp.py:
```diff
@@ -373,6 +373,100 @@ def process_distant_nodes(
+    -------
+    DataFrame
+    """
+    df[new_name] = (df[old_name] <= access_dist).astype("Int64").fillna(0)
```
The type could be `int` or `np.int64` rather than having to parse that string to a type.

gboeing replied:
Come to think of this, why store this with 64-bits if the integer can only take on values of 0 or 1?

carlhiggs replied:
It could be to handle nulls perhaps? (not sure if there should be valid cases of nulls, but if nulls are possible perhaps int64 is required?); eg as per https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html

gboeing replied:
We're already calling `fillna`, so there should be no nulls. We can swap the `astype` and `fillna` order of operations if needed.

carlhiggs replied:
As per the below comments that 'old_names' isn't required, the function could be simplified as:
```
def binary_access_score(df, distances, access_dist):
    """
    Calculate accessibiity score using binary measure: 1 if access <= access_dist, 0 otherwise
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
         list of distances field names
    access_dist: int
        access distance threshold
    Returns
    -------
    DataFrame
    """
    scores = (df[distances] <= access_dist).astype("Int64").fillna(0)
    return scores
```
Using the dataframe constructed below to demonstrate functionality of the soft- and cumulative Gaussian functions, this looks like:
```
df[[f'bin_{x}' for x in distances]] = binary_access_score(df, distances, access_dist)
df
     x1    x2   soft_x1   soft_x2    cum_x1    cum_x2  bin_x1  bin_x2
0   100  1000  0.982014  0.006693  1.000000  0.082085       1       0
1   200   900  0.952574  0.017986  1.000000  0.201897       1       0
2   300   800  0.880797  0.047426  1.000000  0.406570       1       0
3   400   700  0.731059  0.119203  1.000000  0.670320       1       0
4   500   600  0.500000  0.268941  1.000000  0.904837       1       0
5   600   500  0.268941  0.500000  0.904837  1.000000       0       1
6   700   400  0.119203  0.731059  0.670320  1.000000       0       1
7   800   300  0.047426  0.880797  0.406570  1.000000       0       1
8   900   200  0.017986  0.952574  0.201897  1.000000       0       1
9  1000   100  0.006693  0.982014  0.082085  1.000000       0       1
```

carlhiggs commented on process/sp.py:
```diff
@@ -277,14 +277,15 @@
-    samplePointsData[binary_names] = (samplePointsData[distance_names] <= parameters['accessibility_distance']) \
-                                           .astype("Int64").fillna(0)
-    
+    samplePointsData[binary_names] = ssp.binary_access_score(samplePointsData, binary_names, distance_names, parameters['accessibility_distance'])
```
Rather than bake the binary access score function into the code, perhaps we should make the choice user configurable in the setup (ie to address reviewer request)?  Then when the code gets here it checks for the configured formula type (perhaps binary is default), and then executes it along with any associated parameters.  For example, could be ifelse type logic responding to setup parameters.  What do you think?

shiqin-liu replied:
I thought of this too, but I just went for the easy route. I suppose users could just draw from the module for other methods, while keeping our workflow as is with the binary default. But I am open to the idea with ifelse option as well. Would you mind provide me with the code you have in mind? Thank you!

carlhiggs replied:
Maybe what could work is if in the setup code there is a variable defined like
```
threshold_function = 'binary'
```
Then, in sp.py this is used like in the following (using a mock up samplePointsData frame):
```
# establish mock variables and parameters
samplePointsData = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
distance_names = ['x1','x2'] 
access_names = [f'access_{x}' for x in distance_names]
parameters = {'accessibility_distance':500}

# create a dictionary of the available threshold functions -- 
# this could be done in setup_sp.py and called as ssp.threshold_functions[config[threshold_function]]
threshold_functions = {
    'binary':binary_access_score,
    'soft':soft_access_score,
    'cumulative':cumulative_gaussian_access_score
}
# define the access score variables
samplePointsData[access_names] = threshold_functions[threshold_function](samplePointsData, distance_names, parameters['accessibility_distance'])
```

in actual use, I think the line in sp.py would look like,
```
# define the access score variables
samplePointsData[access_names] = ssp.threshold_functions[config[threshold_function]](
    samplePointsData, distance_names, parameters['accessibility_distance']
)
```

It seems to work in this example:
```
>>> samplePointsData = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
>>> distance_names = ['x1','x2']
>>> access_names = [f'access_{x}' for x in distance_names]
>>> parameters = {'accessibility_distance':500}
>>>
>>> threshold_functions = {
...     'binary':binary_access_score,
...     'soft':soft_access_score,
...     'cumulative':cumulative_gaussian_access_score
... }
>>>
>>> samplePointsData[access_names] = threshold_functions['binary'](samplePointsData, distance_names, parameters['accessibility_distance'])
>>>
>>> samplePointsData
       x1    x2  access_x1  access_x2
0   100.0  1000          1          0
1     NaN   900          0          0
2   300.0   800          1          0
3   400.0   700          1          0
4   500.0   600          1          0
5   600.0   500          0          1
6   700.0   400          0          1
7   800.0   300          0          1
8   900.0   200          0          1
9  1000.0   100          0          1
>>>
```

shiqin-liu replied:
While I am working on this change, I felt like it seems more straightforward to just change the functions from sp.py, than  in the setup config code. 

It took me a while to understand the below:
```
ssp.threshold_functions[config[threshold_function]](
    samplePointsData, distance_names, parameters['accessibility_distance']
)
```

So I am thinking may be we should just keep it simple:
```
# create binary access scores evaluated against accessibility distance
# Options for distance decay accessibility scores are available in setup_sp.py module
access_score_names = [f"{x.replace('nearest_node','access')}_score" for x in distance_names]
samplePointsData[access_score_names] = ssp.binary_access_score(samplePointsData, distance_names, parameters['accessibility_distance'])
```

Our goal is to give users the alternatives of distance decay functions, but we still want to state that binary is the preferable option. I guess to keep the simple binary version in the main process would work for our advantages. Just want to share a thought, but I am open to the idea of config setup if anyone think that would be a better option!

carlhiggs replied:
If you think that's best for submitting the article sooner than later, that's fine with me.  Alternately, if you thought it would be good I could do a commit with code for the setup next week (ie use of binary function as a default, matching our use case, but optional parameterisation to use the other functions external to the main code).  I understand if you want to keep things straightforward and lock in the changes now though.

carlhiggs commented on process/sp.py:
```diff
@@ -277,14 +277,15 @@
+
     # create binary distances evaluated against accessibility distance
+    # Options for distance decay accessibility scores are available in setup_sp.py module
     binary_names = [f"{x.replace('nearest_node','access')}_binary" for x in distance_names]
```
We should change the 'binary_names' variable and suffix to something like 'access_score' so it makes sense in a generalised context where a user can configure usage of a custom function

carlhiggs commented on process/setup_sp.py:
```diff
@@ -373,6 +373,100 @@ def process_distant_nodes(
+
+#Soft threshold access score
+#Higgs, C., Badland, H., Simons, K. et al. The Urban Liveability Index
+def soft_access_score(df, new_name, old_name, access_dist, k):
```
Two changes I would make to these functions are
1) No need to provide the 'new name' as this is defined externally to the function, we just need to return a dataframe with some values
2) It would be good to provide default values (if possibly based on reference) to suggest sensible 'k' given 'access_dist'

So, this function could be re-written as
```
#Soft threshold access score
#Higgs, C., Badland, H., Simons, K. et al. The Urban Liveability Index
def soft_access_score(df, distances, access_dist=500, k=5):
    """
    Calculate accessibiity score using soft threshold approach:
    1 / (1+ e ^(k *((dist-access_dist)/access_dist)))
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
        list of field names for distance records
    access_dist: int
        access distance threshold
    k: int
        the slope of decay
    Returns
    -------
    DataFrame
    """
    scores = (1 / (1+numpy.exp(k * ((df[distances]-access_dist) / access_dist)))).astype(float).fillna(0)
    return(scores)
```

An example run through of this shows it appears to work --- (I think this removal of 'old_name' will make the cumulative Gaussian function in particular simpler, avoiding the loop)

```
df = pandas.DataFrame({'x1':[100,200,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})

distances = ['x1','x2'] 
access_dist = 500
k = 5

df[[f'soft_{x}' for x in distances]] =  soft_access_score(df, distances, access_dist=500, k=5)
df
```
and this gives
```
     x1    x2   soft_x1   soft_x2
0   100  1000  0.982014  0.006693
1   200   900  0.952574  0.017986
2   300   800  0.880797  0.047426
3   400   700  0.731059  0.119203
4   500   600  0.500000  0.268941
5   600   500  0.268941  0.500000
6   700   400  0.119203  0.731059
7   800   300  0.047426  0.880797
8   900   200  0.017986  0.952574
9  1000   100  0.006693  0.982014
```

shiqin-liu replied:
Thanks, Carl. This seems more straightforward. I will make the two changes you proposed.

carlhiggs commented on process/setup_sp.py:
```diff
@@ -373,6 +373,100 @@ def process_distant_nodes(
+#Cumulative-Gaussian
+#Reference: Vale, D. S., & Pereira, M. (2017).
+#The influence of the impedance function on gravity-based pedestrian accessibility measures
+def Cumulative_Gaussian_access_score(df, new_name, old_name, access_dist, k):
```
I think we should provide some default values to guide the user as to possible sensible parameterisation of k; e.g.
def Cumulative_Gaussian_access_score(df, new_name, old_name, access_dist=500, k=5)
I'm not sure if k=5 is sensible for this formula for a distance of 500 metres, but if it might be handy having some kind of suggestion of an appropriate decay for 500 metres (ie. if we were to use it for our case study, what would be appropriate?)

I'm not sure why it is necessary to loop through columns in this function, but I think there is a problem in the logic --- in my tests, it seems to replace all values of the 'old name' variable in the input dataframe with 0 as a result of running the function. I am not certain how best to parameterise the decay, but so far (and possibly because all inputs are replaced as zero), the derived scores all evaluate as 1.0 for me? I had a look to see if I can refactor, but perhaps I'm confused; any thoughts?

Here is my test of the basic case where a distance of 700m was recorded relative to a target threshold of 500; intuitively, this should score something below 1 and probably above 0.5.  I struggled to get it to approach that without using extreme values for k; I found 100,000 worked as a usable value in the end though:

```
for k in [.0001,.001,.01,.1,1,10,100,1000,10000,100000,1000000]:
...   print(f"k={k:10}:\t{numpy.exp(-1 * (((700-500)**2) / k))}")
```
...
k=    0.0001:   0.0
k=     0.001:   0.0
k=      0.01:   0.0
k=       0.1:   0.0
k=         1:   0.0
k=        10:   0.0
k=       100:   1.9151695967140057e-174
k=      1000:   4.248354255291589e-18
k=     10000:   0.01831563888873418
k=    100000:   0.6703200460356393
k=   1000000:   0.9607894391523232

so, using k=100,000, perhaps this function could be an alternative, avoiding the looping (which seemed to be implemented in a way that was destructive of the input source in my tests)

```
def Cumulative_Gaussian_access_score(df, distances, access_dist=500, k=100000):
    """
    Calculate accessibiity score using Cumulative-Gaussian approach:
    1 if d <= access_dist ; otherwise, e ^(-1 *((d^2)/k)) if d > access_dist
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
        list of field names for distance records
    access_dist: int
        access distance threshold
    k: int
        the slope of decay
    Returns
    -------
    DataFrame
    """
    df1 = df[distances].copy()
    df1[df1<=access_dist] = 1
    df1[df1>access_dist] = numpy.exp(-1 * (((df1[df1>access_dist]-access_dist)**2) / k))
    df1 = df1.astype(float).fillna(0)
    return(df1)
```

An example of implementation is

```
k=100000
df[[f'cum_{x}' for x in distances]] = Cumulative_Gaussian_access_score(df, distances, access_dist, k)
df
```

This yields the following, adding onto the dataframe with soft threshold inputs defined earlier,
```

     x1    x2   soft_x1   soft_x2    cum_x1    cum_x2
0   100  1000  0.982014  0.006693  1.000000  0.082085
1   200   900  0.952574  0.017986  1.000000  0.201897
2   300   800  0.880797  0.047426  1.000000  0.406570
3   400   700  0.731059  0.119203  1.000000  0.670320
4   500   600  0.500000  0.268941  1.000000  0.904837
5   600   500  0.268941  0.500000  0.904837  1.000000
6   700   400  0.119203  0.731059  0.670320  1.000000
7   800   300  0.047426  0.880797  0.406570  1.000000
8   900   200  0.017986  0.952574  0.201897  1.000000
9  1000   100  0.006693  0.982014  0.082085  1.000000

```

shiqin-liu replied:
> I'm not sure why it is necessary to loop through columns in this function, but I think there is a problem in the logic --- in my tests, it seems to replace all values of the 'old name' variable in the input dataframe with 0 as a result of running the function. 

I used for loop to loop through each column, as I remember in my previous test, I was not able to produce a boolean mask for the entire dataframe. Within the for loop, It first creates new columns of 'new name' variables with value 0, then proceed with calculation, I suppose it should not replace the 'old name' variable with 0"


> Here is my test of the basic case where a distance of 700m was recorded relative to a target threshold of 500; intuitively, this should score something below 1 and probably above 0.5. I struggled to get it to approach that without using extreme values for k; I found 100,000 worked as a usable value in the end though:

Thanks for checking on the parameters, I think it is a good idea to keep default parameters, for this Cumulative_Gaussian function, I would refer to Vale and Pereira (2016) https://journals.sagepub.com/doi/10.1177/0265813516641685; they suggested use k=129,842 for accessibility of 500m, perhaps we could this number as default parameters?



I just tested your functions using the sample point dataframe in Vic, but it returns me with the error: 

```
c_m_names = [f"{x.replace('nearest_node','access')}_c_m" for x in distance_names]
samplePointsData[c_m_names] = Cumulative_Gaussian_access_score(samplePointsData, distance_names, parameters['accessibility_distance'])
samplePointsData[c_m_names]
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-135-f909b77ceb1c> in <module>()
----> 1 samplePointsData[[f'cum_{x}' for x in distances]] = Cumulative_Gaussian_access_score(samplePointsData, distance_names, access_dist, k)

<ipython-input-127-182d031e8ddb> in Cumulative_Gaussian_access_score(df, distances, access_dist, k)
     18     """
     19     df1 = df[distances].copy()
---> 20     df1[df1<=access_dist] = 1
     21     df1[df1>access_dist] = numpy.exp(-1 * (((df1[df1>access_dist]-access_dist)**2) / k))
     22     df1 = df1.astype(float).fillna(0)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py in __setitem__(self, key, value)
   3037 
   3038         if isinstance(key, DataFrame) or getattr(key, "ndim", None) == 2:
-> 3039             self._setitem_frame(key, value)
   3040         elif isinstance(key, (Series, np.ndarray, list, Index)):
   3041             self._setitem_array(key, value)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py in _setitem_frame(self, key, value)
   3086         if key.size and not is_bool_dtype(key.values):
   3087             raise TypeError(
-> 3088                 "Must pass DataFrame or 2-d ndarray with boolean values only"
   3089             )
   3090 

TypeError: Must pass DataFrame or 2-d ndarray with boolean values only
```

carlhiggs replied:
Hmm.... that's odd; I can't check with the actual data right now (some IT issues with RMIT, so can't log in to work computer), but using what should be equivalent test data in the code below it appears to work,
```
>>> def cumulative_gaussian_access_score(df, distances, threshold=500, k=129842):
...     """
...     Calculate accessibiity score using Cumulative-Gaussian approach:
...     1 if d <= threshold ; otherwise, e ^(-1 *((d^2)/k)) if d > threshold
...     Parameters
...     ----------
...     df: DataFrame
...         DataFrame with origin-destination distances
...     distances: list
...         list of field names for distance records
...     threshold: int
...         access distance threshold
...     k: int
...         the slope of decay
...     Returns
...     -------
...     DataFrame
...     """
...     distances = df[distances].copy()
...     distances[distances<=threshold] = 1
...     distances[distances>threshold] = numpy.exp(-1 * (((distances[distances>threshold]-threshold)**2) / k))
...     scores = distances.fillna(0)
...     return scores
...
>>>
>>>
>>> df = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
>>>
>>> distances = ['x1','x2']
>>> df[[f'access_{x}' for x in distances]] = cumulative_gaussian_access_score(df,distances,threshold)
>>> df
       x1    x2  access_x1  access_x2
0   100.0  1000   1.000000   0.145815
1     NaN   900   0.000000   0.291631
2   300.0   800   1.000000   0.499999
3   400.0   700   1.000000   0.734866
4   500.0   600   1.000000   0.925874
5   600.0   500   0.925874   1.000000
6   700.0   400   0.734866   1.000000
7   800.0   300   0.499999   1.000000
8   900.0   200   0.291631   1.000000
9  1000.0   100   0.145815   1.000000
```

shiqin-liu replied:
The test data also work for me too, but still not working for the actual data, I am checking to see what would be the issue there...

carlhiggs replied:
That's really strange... I'm not really sure why it works in the example but fails in the data... but you're right, it has to work in the data!  If you step through line by line is it more clear what isn't working with the Vic data? e.g. for the example which works (with Pandas 1.2.2, Python 3.8),
```
import pandas as pd
import numpy

df = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
distances = ['x1','x2'] 
threshold=500

distances = df[distances].copy()
distances[distances<=threshold] = 1
distances[distances>threshold] = numpy.exp(-1 * (((distances[distances>threshold]-threshold)**2) / k))
scores = distances.fillna(0)
```
I'll try to check it out with the actual data and code later in the week.

shiqin-liu replied:
My best guess is the problem with the missing values, it seems to work when I do:
```
df1 = samplePointsData[distance_names].copy()
mask = (df1<=access_dist).fillna(False) 
mask = mask.astype(bool) # have to change this to bool, otherwise it return me "object" data type and not work as boolean mask
df1[mask]
```

I will check back tomorrow on this, and will keep you updated if all work :)

shiqin-liu replied:
I think I found a fix to the problem, this is related to the data types. Now the function reads:
```
#Cumulative-Gaussian
#Reference: Vale, D. S., & Pereira, M. (2017).
#The influence of the impedance function on gravity-based pedestrian accessibility measures

def cumulative_gaussian_access_score(df, distance_names, threshold=500, k=129842):
    """
    Calculate accessibiity score using Cumulative-Gaussian approach:
    1 if d <= access_dist ; otherwise, e ^(-1 *((d^2)/k)) if d > access_dist

    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distance_names: list
        list of field names for distance records
    threshold: int
        access distance threshold
    k: int
        the slope of decay

    Returns
    -------
    DataFrame
    """
    df1 = df[distance_names].copy()
    df1 = df1.astype(float)
    df1[df1<=threshold] = 1
    df1[df1>threshold] = numpy.exp(-1 * (((df1[df1>threshold]-threshold)**2) / k))
    df1 = df1.fillna(0).astype(float)
    return df1
```
By adding `df1 = df1.astype(float)` seem to resolve the issue, this works when testing with Vic data

carlhiggs reviewed (commented):
Thanks Shirley - I think its a great start; we've discussed a few changes, and I left some suggestions in the issue thread ( https://github.com/global-healthy-liveable-cities/global-indicators/issues/111#issuecomment-783839892 ).  Looking forward to catching up on Friday to discuss this further, Carl

shiqin-liu committed:
modify accessibility functions

- provide default values of k for a given default access distance threshold (comment)
- simplify accessibility score functions without the new names (comment)
- doing fillna(0) prior to casting type to avoid the need of using int64 for a nullable datatype. (comment)
- update cumulative–gaussian function, without the need of for loop (comment)

shiqin-liu commented:
> Thanks Shirley - I think its a great start; we've discussed a few changes, and I left some suggestions in the issue thread ( [#111 (comment)](https://github.com/global-healthy-liveable-cities/global-indicators/issues/111#issuecomment-783839892) ). Looking forward to catching up on Friday to discuss this further, Carl

Thanks Carl, I have addressed some of the comments, and will look into the accessibility name change and the user configurable options in the set up file.

shiqin-liu committed:
amend the access score variable names

change "binary" names to a more general term access "score"

gboeing reviewed (approved):
This looks good to me. I defer to @carlhiggs for final approval. Carl, when you approve, feel free to merge the PR. You and Shirley have admin access on this repo.

carlhiggs commented:
Hi @gboeing and @shiqin-liu,

Sounds good - just so you know, I'll be on leave from today returning
Monday 8 March --- I don't believe I'll be able to review the changes
properly before then. Hope that's okay.

Cheers,

Carl

On Wed, 3 Mar. 2021, 09:02 Geoff Boeing, <notifications@github.com> wrote:

> *@gboeing* approved this pull request.
>
> This looks good to me. I defer to @carlhiggs
> <https://github.com/carlhiggs> for final approval. Carl, when you
> approve, feel free to merge the PR. You and Shirley have admin access on
> this repo.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/global-healthy-liveable-cities/global-indicators/pull/112#pullrequestreview-602313244>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDCCMXLRIPP3XIYJJVPMFLTBVN55ANCNFSM4YBIBVRQ>
> .
>

shiqin-liu commented:
@carlhiggs That's OK. Feel free to merge the PR when you approve, or let me know if you find any issues.

carlhiggs commented on process/setup_config.py:
```diff
@@ -222,24 +222,24 @@
                     }}
             }},
-            # evaluate sum of binary scores, ignoring nulls
+            # evaluate sum of score scores, ignoring nulls
```
```suggestion
            # evaluate sum of access scores, ignoring nulls

```

carlhiggs reviewed (approved):
All looks good to me, thanks Shirley.  I made one trivial suggested change to a comment (score scores to access scores), but the code itself looks all good, implementing the changes discussed.  I think its fine not having the score function directly configurable for now.  Nice work!

shiqin-liu committed:
correct typo
